inductor: avoiding moving constructor to cuda when it would cause h2d sync in index_put_ fallback

bdhirsh opened the pull request:
My attempt at a fix for https://github.com/pytorch/pytorch/issues/130335, see issue for more details / internal xref. Any feedback from inductor folks is appreciated. I attempted to make the move-constructors-to-cuda pass a bit less aggressive by detecting when the movement would incur a H2D sync for `aten.index_put_`. I'm not sure if there are any other ops that inductor falls back to eager on, that may-or-may-not incur a H2D sync if we change any of their inputs from cpu to cuda.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128331
* #128045
* __->__ #130338



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

bdhirsh committed:
inductor: avoiding moving constructor to cuda when it would cause h2d sync in index_put_ fallback

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1015,6 +1015,29 @@ def fused_int_mm_mul(match: Match, mat1, mat2, mat3, out_dtype=None):
+    # Therefore, when inductor sees an index_put_ with byte tensor indices,
+    # it should *not* convert the cpu scalar value into a cuda tensor.
+    any_byte_bool_indices = False
+    indices = node.args[1]
```
nit: use `operator_schemas.normalize_function` instead of directly accessing args

eellison commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -1015,6 +1015,29 @@ def fused_int_mm_mul(match: Match, mat1, mat2, mat3, out_dtype=None):
 
 
+def is_index_put_and_requires_h2d_sync_for_cuda_value(node):
+    if node.target is not torch.ops.aten.index_put.default:
```
do we want this for index_put_ as well ?

bdhirsh replied:
I figured it would just never get hit since functionalization has already run, and reinplacing runs in a later pass. But probably doesn't hurt to add, can add it

eellison reviewed (approved):
Looks good ! Test ?

bdhirsh commented:
> Looks good ! Test ?

hmmm let me know if you have better testing ideas - the idea of a "no h2d sync happened" test feels a bit fragile (e.g. I could have the test run the profiler and parse the output to ensure that `nonzero()` never shows up, but that doesn't feel great).

Alternatively, I could just add an expecttest test where I assert that the final inductor graph properly keeps the scalar-tensor on cpu when the above conditions apply, as a proxy for the H2D sync test (i'll probably default to that)

eellison commented:
@bdhirsh yea sg, that's what i was thinking

bdhirsh committed:
Update on "inductor: avoiding moving constructor to cuda when it would cause h2d sync in index_put_ fallback"

My attempt at a fix for https://github.com/pytorch/pytorch/issues/130335, see issue for more details / internal xref. Any feedback from inductor folks is appreciated. I attempted to make the move-constructors-to-cuda pass a bit less aggressive by detecting when the movement would incur a H2D sync for `aten.index_put_`. I'm not sure if there are any other ops that inductor falls back to eager on, that may-or-may-not incur a H2D sync if we change any of their inputs from cpu to cuda.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "inductor: avoiding moving constructor to cuda when it would cause h2d sync in index_put_ fallback"

My attempt at a fix for https://github.com/pytorch/pytorch/issues/130335, see issue for more details / internal xref. Any feedback from inductor folks is appreciated. I attempted to make the move-constructors-to-cuda pass a bit less aggressive by detecting when the movement would incur a H2D sync for `aten.index_put_`. I'm not sure if there are any other ops that inductor falls back to eager on, that may-or-may-not incur a H2D sync if we change any of their inputs from cpu to cuda.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
land ?

bdhirsh commented:
@pytorchbot merge
